base/array.jl

```diff
@@ -667,6 +687,10 @@ function collect_to!(dest::AbstractArray{T}, itr, offs, st) where T
             return collect_to!(new, itr, i+1, st)
         end
     end
+    i-1 < length(dest) &&
+        throw(ErrorException("iterator returned fewer elements than its declared length"))
+    i-1 > length(dest) &&
```

Comment:
In this case, the `@inbounds dest[i]` above has likely corrupted something or segfaulted. Should probably replace the `while true` with `while i <= lastindex(dest)` and verify here that `iterate(itr, st) === nothing` (or `Base.isdone(itr, st) === true || iterate(itr, st) === nothing `?). Should then fix  #29532.

Reply:
In theory I agree, I don't like this `@inbounds` call which isn't associated with an actual check of the index against the array. But it comes with an additional cost with hurts performance at lot since it disables SIMD (see discussion [above](https://github.com/JuliaLang/julia/pull/29458#discussion_r221740803). I've just tried it:
```julia
# Master
julia> x = Vector{Int}(undef, 1000);

julia> itr = 1:1000;

julia> @btime Base.collect_to!($x, $itr, 1, 0);
  676.981 ns (0 allocations: 0 bytes)

julia> y = copy(x);

julia> @btime Base.collect_to!($x, $y, 1, 1);
  847.618 ns (0 allocations: 0 bytes)

# After change

julia> x = Vector{Int}(undef, 1000);

julia> itr = 1:1000;

julia> @btime Base.collect_to!($x, $itr, 1, 0);

  86.233 ns (0 allocations: 0 bytes)

julia> y = copy(x);

julia> @btime Base.collect_to!($x, $y, 1, 1);
  90.980 ns (0 allocations: 0 bytes)
```

Maybe the compiler could be improved, but for now I guess we can't do this.

Reply:
Ouch, that's bad.